### PR TITLE
Improve AdGuardHome binary-replace recovery and committed-backup cleanup; add tests

### PR DESCRIPTION
**Pull request by jumpsmm7**
### Motivation
- Harden installer handling of interrupted binary replacements so a restored binary and service restart are retried reliably across EXIT and signal paths.
- Introduce a safe, auditable mechanism to remove committed previous-binary backups across process restarts without risking arbitrary file deletion.
- Add automated tests to validate rollback state survivability, cleanup retry behavior, and lifecycle gating for pending committed-backup removal.

### Description
- Added `adguard_install_signal_traps_disable()` and refactored trap disable/enable logic so signal handling can be controlled independently of state reset.
- Implemented restart-pending semantics (`ADGUARD_INSTALL_RESTART_PENDING`) and adjusted `adguard_restore_after_failed_replace()`, `adguard_restart_after_install_abort()` and `on_installer_exit()` to defer and retry service restarts when appropriate.
- Added committed-backup lifecycle helpers: `adguard_committed_binary_cleanup_path_valid()`, `adguard_committed_binary_cleanup_record_write()`, and `adguard_committed_binary_cleanup_retry()` to record, validate, and retry removal of `.AdGuardHome.previous.*` backups safely, including startup gating.
- Improved event-hook transaction marker cleanup and error reporting, and disarmed replacement state before attempting committed-backup removal during install finalization.

### Testing
- Added and ran new tests `tests/installer-binary-rollback-state.sh`, `tests/installer-committed-binary-cleanup.sh`, and `tests/installer-committed-binary-lifecycle.sh`, and updated several existing installer tests to exercise the new control paths.
- Updated `tools/code-quality.sh` to include the new regression checks and executed the suite that exercises `installer` unit tests and integration helpers.
- All new and modified automated tests were executed and passed (no regressions reported).

------
[Codex Task](https://chatgpt.com/codex/cloud/tasks/task_e_6aa98f3bc6ec832993868ff06442d54b)

### CODE REVIEW

File: installer [hunk 1]
```diff
@@ -7047,11 +7158,26 @@ inst_AdGuardHome() {
 		fi
 	fi
 	if ! all_event_scripts_transaction_commit; then
+		all_event_scripts_transaction_rollback || true
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @coderabbitai explain this issue as a prompt for an AI Agent

File: installer [hunk 2]
```diff
@@ -5300,7 +5340,72 @@ adguard_restore_after_failed_replace() {
...
+		rm -f "${CLEANUP_STAGE}"
+		return 1
+	}
+}
+
+adguard_committed_binary_cleanup_retry() {
+	local CLEANUP_MARKER CLEANUP_PATH
+	CLEANUP_MARKER="${BASE_DIR}/.AdGuardHome.binary-cleanup"
+	[ -e "${CLEANUP_MARKER}" ] || return 0
+	[ -f "${CLEANUP_MARKER}" ] && [ ! -L "${CLEANUP_MARKER}" ] || {
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @coderabbitai explain this issue as a prompt for an AI Agent

File: installer [hunk 3]
```diff
@@ -7047,11 +7158,26 @@ inst_AdGuardHome() {
...
 	if ! all_event_scripts_transaction_commit; then
+		all_event_scripts_transaction_rollback || true
 		PTXT "${ERROR} Unable to finalize the event-hook recovery transaction."
 		adguard_recover_after_event_hook_abort "${RESTART_AFTER_ABORT}" 1
 		end_op_message 1 "$1"
 		return 1
 	fi
+	if [ "${ADGUARD_INSTALL_REPLACE_ACTIVE:-0}" = "1" ]; then
+		COMMITTED_OLD_BINARY="${ADGUARD_INSTALL_OLD_BINARY}"
+		adguard_install_signal_traps_disable
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @coderabbitai explain this issue as a prompt for an AI Agent

File: installer [hunk 4]
```diff
@@ -5300,7 +5340,72 @@ adguard_restore_after_failed_replace() {
...
+adguard_committed_binary_cleanup_retry() {
+	local CLEANUP_MARKER CLEANUP_PATH
+	CLEANUP_MARKER="${BASE_DIR}/.AdGuardHome.binary-cleanup"
+	[ -e "${CLEANUP_MARKER}" ] || return 0
+	[ -f "${CLEANUP_MARKER}" ] && [ ! -L "${CLEANUP_MARKER}" ] || {
+		PTXT "${ERROR} Unsafe committed-binary cleanup record retained at ${CLEANUP_MARKER}."
+		return 1
+	}
+	IFS= read -r CLEANUP_PATH <"${CLEANUP_MARKER}" || CLEANUP_PATH=""
+	if sed -n '2p' "${CLEANUP_MARKER}" | grep -q . || ! adguard_committed_binary_cleanup_path_valid "${CLEANUP_PATH}"; then
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @coderabbitai explain this issue as a prompt for an AI Agent

File: installer [hunk 5]
```diff
@@ -7047,11 +7158,26 @@ inst_AdGuardHome() {
...
+		COMMITTED_OLD_BINARY="${ADGUARD_INSTALL_OLD_BINARY}"
+		adguard_install_signal_traps_disable
+		ADGUARD_INSTALL_REPLACE_ACTIVE="0"
+		if adguard_committed_binary_cleanup_record_write "${COMMITTED_OLD_BINARY}"; then
+			adguard_committed_binary_cleanup_retry ||
+				PTXT "${WARNING} Unable to remove the committed AdGuardHome binary backup at ${COMMITTED_OLD_BINARY}."
+		elif rm -f "${COMMITTED_OLD_BINARY}"; then
+			:
+		else
+			PTXT "${ERROR} Unable to create ${BASE_DIR}/.AdGuardHome.binary-cleanup for retained committed binary backup ${COMMITTED_OLD_BINARY}."
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @coderabbitai explain this issue as a prompt for an AI Agent

**Comment by jumpsmm7** [comment]
### 💡 Codex Review

https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/11b789c54e568714b5c94c2da8d8f26e59813216/installer#L5859
**<sub><sub>P2 Badge</sub></sub>  Finalize rollback state before the unknown-mode return**

When a package replacement succeeds but `adguard_install_mode_confirmed` is false, `inst_AdGuardHome` returns success at lines 6962–6964 before reaching the new commit block. Because this line now leaves `ADGUARD_INSTALL_REPLACE_ACTIVE` and `ADGUARD_INSTALL_OLD_BINARY` armed, the top-level EXIT handler restores the previous binary and restarts it, silently undoing an update that was reported as successful and potentially leaving newer integration files paired with the old binary. Finalize the binary transaction before that early return, or treat the path as a failure and perform explicit recovery.

AGENTS.md reference: [AGENTS.md:L253-L258](https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/11b789c54e568714b5c94c2da8d8f26e59813216/AGENTS.md#L253-L258)

---

https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/11b789c54e568714b5c94c2da8d8f26e59813216/installer#L5200
**<sub><sub>P1 Badge</sub></sub>  Preserve pending restart state after failed readiness recovery**

When post-install readiness fails and rollback restores the previous binary but its first restart attempt fails, `adguard_restore_after_failed_replace` correctly sets `ADGUARD_INSTALL_RESTART_PENDING=1`; however, the readiness-failure path then unconditionally calls `adguard_install_abort_trap_disable_preserve_defer` at line 7132, and this newly added reset clears that state. The EXIT handler consequently cannot perform its intended restart retry, leaving a previously running AdGuardHome service down after the failed update. Preserve the pending flag when recovery reports failure, or only disable/reset the abort state after restart recovery succeeds.

AGENTS.md reference: [AGENTS.md:L253-L258](https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/11b789c54e568714b5c94c2da8d8f26e59813216/AGENTS.md#L253-L258)

---

https://github.com/jumpsmm7/Asuswrt-Merlin-AdGuardHome-Installer/blob/11b789c54e568714b5c94c2da8d8f26e59813216/installer#L7177-L7179
**<sub><sub>P2 Badge</sub></sub>  Retain a retry path when both cleanup operations fail**

If writing `.AdGuardHome.binary-cleanup` fails and the fallback `rm -f` also fails, this branch only logs an error, then clears `ADGUARD_INSTALL_OLD_BINARY` and lets the installation report success. Since replacement state was already disarmed and no cleanup marker exists, neither EXIT cleanup nor a later installer invocation can discover or retry the retained `.AdGuardHome.previous.*` file, permanently leaking a potentially large binary on constrained `/opt` storage. Keep an in-memory cleanup path for EXIT retry and do not discard it until the backup is removed or durably recorded.
    

<details> <summary>ℹ️ About Codex in GitHub</summary>
<br/>

[Your team has set up Codex to review pull requests in this repo](https://chatgpt.com/codex/cloud/settings/general). Reviews are triggered when you
- Open a pull request for review
- Mark a draft as ready
- Comment "@codex review".

If Codex has suggestions, it will comment; otherwise it will react with 👍.




Codex can also answer questions or update the PR. Try commenting "@codex address that feedback".
            
</details>

@coderabbitai explain the issues present in this comment as a prompt for an AI Agent

File: tests/installer-committed-binary-lifecycle.sh [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	exit 1
+}
+
+cleanup() {
+	rm -rf "${TEST_ROOT}"
+}
+
+trap cleanup 0
+trap 'cleanup; exit 1' HUP INT TERM
+mkdir -p "${TEST_ROOT}" || fail 'could not create committed-binary lifecycle fixture'
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply tests/installer-committed-binary-lifecycle.sh]: @coderabbitai explain this issue as a prompt for an AI Agent

**Comment by jumpsmm7** [comment]
@codex fix:
```text
Fix a reliability defect in `installer` in the committed AdGuardHome binary-backup cleanup path.

Location:
- Function: `inst_AdGuardHome()`
- Area: the block that handles `ADGUARD_INSTALL_REPLACE_ACTIVE=1` after `all_event_scripts_transaction_commit()` succeeds.

Problem:
`inst_AdGuardHome()` preserves the old AdGuardHome binary at a path such as:

    ${TARG_DIR}/.AdGuardHome.previous.$$

After the installation commits, it tries to write a durable cleanup record with:

    adguard_committed_binary_cleanup_record_write "${COMMITTED_OLD_BINARY}"

If record creation fails, it tries this fallback:

    rm -f "${COMMITTED_OLD_BINARY}"

If both operations fail, the code only prints an error. It then clears the in-memory replacement state through `adguard_install_abort_trap_disable_preserve_defer()` and continues to report a successful installation.

This failure path leaves the committed backup on disk without a valid `.AdGuardHome.binary-cleanup` record. EXIT cleanup and the next installer startup cannot discover and retry deletion of that backup.

Required behavior:
1. Preserve durable cleanup state whenever the old binary remains after commit.
2. Do not report successful finalization when both cleanup-record creation and direct backup deletion fail.
3. Return a nonzero status from `inst_AdGuardHome()` for the double-failure path.
4. Keep the committed installation and running AdGuardHome service intact. Do not roll back the successful binary replacement only because cleanup of the old backup failed.
5. Do not clear recovery state before the failure result is returned unless an equivalent durable record exists.
6. Keep signal-trap behavior safe. A signal after the replacement commits must not restore the old binary over the committed binary.
7. Use the existing validated cleanup-record mechanism where possible. Do not create an unvalidated marker format.

Suggested implementation approach:
- Keep `adguard_install_signal_traps_disable` after the replacement commit.
- Try `adguard_committed_binary_cleanup_record_write`.
- If record creation succeeds, call `adguard_committed_binary_cleanup_retry`.
  - A retry failure may return failure while retaining the valid record for future startup retry.
- If record creation fails, attempt direct `rm -f` of the backup.
  - If direct deletion succeeds, continue normally.
  - If direct deletion also fails, print a clear error, preserve safe state as applicable, call `end_op_message 1 "$1"` or return a nonzero status through the existing error path, and do not print successful orchestration messages.

Add regression tests for:
1. Cleanup-record creation failure with direct backup deletion success.
   - Expected result: success and no retained backup.
2. Cleanup-record creation failure with direct backup deletion failure.
   - Expected result: nonzero operation status.
   - Expected result: no successful completion message.
   - Expected result: the committed AdGuardHome binary remains installed.
   - Expected result: the old backup is not silently forgotten.
3. Cleanup-record creation success with cleanup retry failure.
   - Expected result: nonzero status or clearly documented retained-cleanup behavior.
   - Expected result: the valid cleanup record remains for EXIT or startup retry.
4. A later startup retry after a valid retained record.
   - Expected result: the backup and record are removed when deletion becomes possible.

Do not make unrelated refactors.
Maintain POSIX `/bin/sh` and BusyBox v1.25.1 compatibility.

Fix a signal-handling correctness defect in the `installer` shell script.

Context:
`inst_AdGuardHome()` replaces the AdGuardHome binary through a rollback-capable transaction. It preserves the prior binary in `ADGUARD_INSTALL_OLD_BINARY` and marks the replacement active with `ADGUARD_INSTALL_REPLACE_ACTIVE=1`.

Later, `inst_AdGuardHome()` commits the event-hook transaction and finalizes the NVRAM transaction. At that point, the upgrade is committed. The new configuration, event hooks, and router state must remain paired with the new binary.

Defect:
A termination signal can arrive after event-hook and NVRAM commit, but before the binary-replacement state is disarmed or converted to committed-backup cleanup state.

If this occurs, EXIT cleanup or the install signal handler can still interpret:
- `ADGUARD_INSTALL_REPLACE_ACTIVE=1`
- `ADGUARD_INSTALL_OLD_BINARY=<previous binary>`

as an uncommitted replacement. The cleanup path can then restore the old binary. This leaves the old binary with newly committed configuration and router state.

Required changes:
1. Identify the exact post-commit section in `inst_AdGuardHome()` after:
   - `finalize_dns_environment` or `nvram_transaction_finalize_setup_pair`
   - `finalize_pending_mode_migration`
   - `all_event_scripts_transaction_commit`

2. Remove any interval where signal handlers use rollback-capable replacement state after these commit points.

3. Add an explicit committed-finalization transition:
   - Disable or replace the rollback signal handler only after replacement state is no longer rollback-eligible.
   - Copy `ADGUARD_INSTALL_OLD_BINARY` to a local committed-backup variable.
   - Set `ADGUARD_INSTALL_REPLACE_ACTIVE=0`.
   - Clear `ADGUARD_INSTALL_OLD_BINARY`.
   - Ensure a later EXIT handler or signal handler cannot call `adguard_restore_after_failed_replace()` for this committed replacement.

4. Treat the old binary only as a committed backup after the transition:
   - Write a validated persistent cleanup record before removing the backup when possible.
   - Retry committed-backup cleanup through `adguard_committed_binary_cleanup_retry`.
   - If cleanup fails, retain the backup and cleanup record for startup retry.
   - Do not roll back the committed binary replacement when cleanup fails.

5. Verify all relevant cleanup paths:
   - `adguard_install_abort_on_signal`
   - `on_installer_exit`
   - `adguard_restart_after_install_abort`
   - `adguard_restore_after_failed_replace`
   - startup recovery that calls `adguard_committed_binary_cleanup_retry`

6. Preserve these invariants:
   - Before commit, interruption restores the prior binary and prior service state.
   - After commit, interruption never restores the prior binary.
   - After commit, a retained `.AdGuardHome.previous.*` file is only a cleanup artifact.
   - A cleanup failure cannot make the new binary rollback-eligible.
   - The service remains recoverable if a signal occurs during committed-backup cleanup.

7. Add regression tests that simulate a signal or EXIT cleanup at the post-commit boundary. The tests must prove that:
   - rollback is not called after commit;
   - the new binary remains installed;
   - the old binary is retained only through the cleanup record when deletion fails; and
   - a later startup or EXIT retry removes the backup and record when deletion succeeds.

Keep the patch minimal. Maintain POSIX `/bin/sh` compatibility for BusyBox ash.

Fix the dangling cleanup-marker bypass in the `installer` script.

Affected function:
- `adguard_committed_binary_cleanup_retry()`

Problem:
- The function currently treats `.AdGuardHome.binary-cleanup` as absent when `[ -e "${CLEANUP_MARKER}" ]` is false.
- For a dangling symlink, `test -e` returns false but `test -L` returns true.
- The function therefore returns success before it reaches the existing unsafe-marker validation.
- Startup then continues with an unsafe cleanup marker in place.
- A later committed binary replacement cannot publish a new cleanup record because `adguard_committed_binary_cleanup_record_write()` correctly rejects an existing symlink.

Required change:
1. Treat the cleanup marker as absent only when it is neither an existing filesystem object nor a symlink.
2. Change the early-return condition in `adguard_committed_binary_cleanup_retry()` so it returns success only when both checks are false:
   - `[ ! -e "${CLEANUP_MARKER}" ]`
   - `[ ! -L "${CLEANUP_MARKER}" ]`
3. Do not remove or follow a dangling symlink.
4. Let a dangling symlink reach the existing unsafe-record error path.
5. Preserve the current behavior for:
   - no cleanup marker,
   - a valid regular-file cleanup marker,
   - malformed cleanup records,
   - non-symlink unsafe marker types.

Add or update regression tests:
1. Create `.AdGuardHome.binary-cleanup` as a dangling symlink.
2. Call `adguard_committed_binary_cleanup_retry()`.
3. Verify that it fails.
4. Verify that the symlink remains unchanged.
5. Verify that startup dispatch does not continue when this helper fails.
6. Verify that a truly absent marker still returns success.

Run the relevant shell regression tests and the repository code-quality check. Do not change unrelated cleanup or replacement logic.

Fix a reliability defect in the `installer` shell script.

Problem:
In `inst_AdGuardHome()`, the failure branch for
`all_event_scripts_transaction_commit()` runs:

    all_event_scripts_transaction_rollback || true

The `|| true` suppresses a failed event-hook rollback. The code then calls
`adguard_recover_after_event_hook_abort()`.

During a mode migration, `adguard_recover_after_event_hook_abort()` can
successfully roll back the mode migration and call
`all_event_scripts_transaction_detach_after_mode_rollback()`.

That detach helper removes:
- `EVENT_SCRIPTS_ACTIVE_SNAPSHOT`
- `${BASE_DIR}/.AdGuardHome.event-hooks-recovery`
- the event-hook snapshot directory

If `all_event_scripts_transaction_rollback()` failed, the snapshot is still
required. It contains the recovery state for a later EXIT or startup retry.
Deleting it can leave the router with partially restored event-hook files and
no recovery copy.

Required change:
1. In `inst_AdGuardHome()`, preserve the exit status from
   `all_event_scripts_transaction_rollback()` in the
   `all_event_scripts_transaction_commit()` failure path.
2. Pass the rollback result to `adguard_recover_after_event_hook_abort()`, or
   otherwise make that helper aware that the event-hook rollback failed.
3. If the event-hook rollback failed, do not call
   `all_event_scripts_transaction_detach_after_mode_rollback()`.
4. Keep `EVENT_SCRIPTS_ACTIVE_SNAPSHOT` and
   `${BASE_DIR}/.AdGuardHome.event-hooks-recovery` intact when the rollback
   fails.
5. Continue independent recovery steps where safe:
   - restore the binary and service state;
   - restore NVRAM state;
   - roll back mode-migration state.
6. Do not delete the event-hook snapshot until
   `all_event_scripts_transaction_rollback()` succeeds.
7. Keep the existing behavior unchanged when event-hook rollback succeeds.

Affected code:
- `installer`: `inst_AdGuardHome()`
- `installer`: `adguard_recover_after_event_hook_abort()`
- `installer`: `all_event_scripts_transaction_detach_after_mode_rollback()`
- `installer`: `all_event_scripts_transaction_rollback()`

Add a regression test that simulates all of these conditions:
- `all_event_scripts_transaction_commit()` fails;
- `all_event_scripts_transaction_rollback()` fails;
- a mode migration is active;
- mode rollback itself succeeds.

Verify that:
- the event-hook recovery marker remains;
- the event-hook snapshot remains;
- `EVENT_SCRIPTS_ACTIVE_SNAPSHOT` remains set;
- startup recovery can retry the event-hook rollback;
- independent binary, service, and NVRAM recovery still runs where applicable.

Use POSIX `/bin/sh` syntax. Do not introduce Bash-only syntax or new runtime
dependencies. Use the smallest change that preserves the recovery state.

Review and fix the test-fixture setup in `tests/installer-committed-binary-lifecycle.sh`.

Problem:
`TEST_ROOT` uses a predictable path:

    ${TMPDIR:-/tmp}/installer-binary-cleanup-lifecycle.$$

The script creates this path with:

    mkdir -p "${TEST_ROOT}"

If another user creates this path first, including as a symlink, `mkdir -p` can succeed. Later test writes can then follow that path and overwrite files outside the intended fixture directory. This is relevant when the test runs with elevated privileges or in a shared temporary directory.

Required change:
1. Create `TEST_ROOT` exclusively. Do not use `mkdir -p`.
2. Fail if `TEST_ROOT` already exists.
3. Restrict permissions on the new fixture directory.
4. Preserve the existing failure message:
   `could not create committed-binary lifecycle fixture`
5. Do not change the test behavior after successful directory creation.

Use this setup:

    (umask 077 && mkdir "${TEST_ROOT}") || fail 'could not create committed-binary lifecycle fixture'

Validation:
- Run `sh -n tests/installer-committed-binary-lifecycle.sh`.
- Confirm that an existing `TEST_ROOT` causes an immediate failure.
- Confirm that fixture file writes occur only after exclusive directory creation succeeds.

Review and fix three binary-replacement transaction defects in `installer`.

Do not change unrelated installer behavior. Preserve the existing transaction-marker and cleanup-record design. Add regression tests for each failure path.

1. Finalize the replacement transaction before the unknown-mode success return.

Location:
- `inst_AdGuardHome`, near the early successful return where `adguard_install_mode_confirmed` is false.

Failure mode:
- A package replacement can succeed while `adguard_install_mode_confirmed` is false.
- The function returns success before it reaches the normal commit block.
- `ADGUARD_INSTALL_REPLACE_ACTIVE` and `ADGUARD_INSTALL_OLD_BINARY` remain armed.
- The top-level `EXIT` handler then treats the successful update as an aborted replacement.
- The handler restores the previous binary and restarts it.
- The installer reports success even though it has reverted the installed binary.
- New integration files can remain installed with the old binary.

Required change:
- Before this early success return, finalize the binary replacement transaction using the same state transitions as the normal successful commit path.
- Alternatively, treat this path as a failure and run explicit recovery.
- Do not allow a successful return with rollback state still armed.

Required test:
- Simulate a successful replacement with `adguard_install_mode_confirmed=false`.
- Verify that the `EXIT` handler does not restore the old binary.
- Verify that the new binary remains installed.
- Verify that transaction state is disarmed or committed before return.

2. Preserve restart retry state after failed readiness recovery.

Location:
- The post-install readiness-failure path in `inst_AdGuardHome`, near the call to `adguard_install_abort_trap_disable_preserve_defer`.
- `adguard_restore_after_failed_replace`.

Failure mode:
- Readiness fails after a replacement.
- Rollback restores the old binary.
- The first restart of the old service fails.
- `adguard_restore_after_failed_replace` sets `ADGUARD_INSTALL_RESTART_PENDING=1`.
- The readiness-failure path then calls `adguard_install_abort_trap_disable_preserve_defer`.
- That reset clears `ADGUARD_INSTALL_RESTART_PENDING`.
- The `EXIT` handler cannot retry the restart.
- A service that was running before the update can remain stopped.

Required change:
- Preserve `ADGUARD_INSTALL_RESTART_PENDING` when rollback recovery reports a failed restart.
- Only clear abort and restart-retry state after restart recovery succeeds.
- Ensure the `EXIT` handler retains enough state to retry the restart.

Required test:
- Simulate readiness failure after binary replacement.
- Simulate successful binary restoration and failed first restart.
- Verify that `ADGUARD_INSTALL_RESTART_PENDING` remains set after the readiness-failure path.
- Verify that exit cleanup attempts the restart retry.

3. Retain a cleanup retry path when both committed-backup cleanup operations fail.

Location:
- The committed-backup cleanup path near the branch that writes `.AdGuardHome.binary-cleanup` and falls back to `rm -f`.

Failure mode:
- The replacement transaction is committed.
- Writing `.AdGuardHome.binary-cleanup` fails.
- The fallback removal of `.AdGuardHome.previous.*` also fails.
- The code only logs an error and clears `ADGUARD_INSTALL_OLD_BINARY`.
- The replacement state is already disarmed.
- No cleanup record exists for a later installer invocation.
- The current process and future processes cannot discover the retained backup.
- The backup can remain permanently and consume limited `/opt` storage.

Required change:
- Do not discard the previous-binary path until the backup is removed or its cleanup record is stored durably.
- When cleanup-record creation and immediate removal both fail, retain an in-memory cleanup path for the `EXIT` handler to retry.
- Keep the cleanup retry state until removal succeeds or durable state exists for startup cleanup.
- Report the failure clearly without reporting cleanup as complete.

Required test:
- Simulate failure to write `.AdGuardHome.binary-cleanup`.
- Simulate failure of the immediate `rm -f` fallback.
- Verify that the previous-binary path remains available for exit-time cleanup retry.
- Verify that a later successful retry removes the backup and clears the retained state.

Validation:
- Run the existing installer lifecycle and binary-replacement regression tests.
- Add the new tests to `tools/code-quality.sh` if that script is the repository’s required test entry point.
- Confirm that successful replacements cannot be rolled back by `EXIT`.
- Confirm that failed rollback restarts remain retryable.
- Confirm that failed committed-backup cleanup remains retryable.
```

File: installer [hunk 7]
```diff
@@ -5421,6 +5435,28 @@ adguard_committed_binary_cleanup_retry() {
...
 }
 
+# adguard_committed_binary_cleanup_finalize disarms rollback and removes or durably records the committed backup in $1.
+adguard_committed_binary_cleanup_finalize() {
+	local COMMITTED_OLD_BINARY
+	COMMITTED_OLD_BINARY="$1"
+	ADGUARD_INSTALL_REPLACE_ACTIVE="0"
+	ADGUARD_INSTALL_OLD_BINARY=""
+	ADGUARD_INSTALL_COMMIT_PENDING="0"
+	adguard_install_signal_traps_disable
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @coderabbitai explain this issue as a prompt for an AI Agent

File: installer [hunk 8]
```diff
@@ -5421,6 +5435,28 @@ adguard_committed_binary_cleanup_retry() {
...
+	else
+		PTXT "${ERROR} Unable to create ${BASE_DIR}/.AdGuardHome.binary-cleanup or remove retained committed binary backup ${COMMITTED_OLD_BINARY}."
+		return 1
+	fi
+	return 0
+}
+
 adguard_restore_after_failed_directory_restore() {
 	local COMMAND_LINK ROLLBACK_DIR STAGE_DIR TARGET_DIR TARGET_INSTALLED WAS_RUNNING
 	ROLLBACK_DIR="$1"
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer:5467]: @coderabbitai explain this issue as a prompt for an AI Agent

File: installer [hunk 9]
```diff
@@ -5421,6 +5435,33 @@ adguard_committed_binary_cleanup_retry() {
 	return 0
 }
 
+# adguard_committed_binary_cleanup_finalize disarms rollback and removes or durably records the committed backup in $1.
+adguard_committed_binary_cleanup_finalize() {
+	local COMMITTED_OLD_BINARY
+	COMMITTED_OLD_BINARY="$1"
+	[ -n "${COMMITTED_OLD_BINARY}" ] || return 0
+	ADGUARD_COMMITTED_BINARY_CLEANUP_PENDING="${COMMITTED_OLD_BINARY}"
+	ADGUARD_INSTALL_COMMIT_PENDING="1"
```
[the remark this answers is not in the document]
Comment (jumpsmm7) [reply installer]: @coderabbitai explain this issue as a prompt for an AI Agent
